pkg/core/statesync/neotest_test.go

```diff
@@ -475,9 +685,113 @@ func TestStateSyncModule_RestoreBasicChain(t *testing.T) {
 		require.False(t, haveItems)
 	}
 	t.Run("source node is archive", func(t *testing.T) {
-		check(t, false)
+		check(t, false, false)
 	})
 	t.Run("source node is light with GC", func(t *testing.T) {
-		check(t, true)
+		check(t, true, false)
+	})
+	t.Run("ContractStorageBased mode", func(t *testing.T) {
+		check(t, false, true)
 	})
 }
+
+func TestStateSyncModule_SetOnStageChanged(t *testing.T) {
+	const (
+		stateSyncInterval = 2
+		maxTraceable      = 3
+	)
+	spoutCfg := func(c *config.Blockchain) {
+		c.StateRootInHeader = true
+		c.StateSyncInterval = stateSyncInterval
+		c.MaxTraceableBlocks = maxTraceable
+	}
+	bcSpout, vals, comm := chain.NewMultiWithCustomConfig(t, spoutCfg)
+	e := neotest.NewExecutor(t, bcSpout, vals, comm)
+	for range 2*stateSyncInterval + maxTraceable + 2 {
+		e.AddNewBlock(t)
+	}
+
+	mptCfg := func(c *config.Blockchain) {
+		spoutCfg(c)
+		c.P2PStateExchangeExtensions = true
+		c.Ledger.KeepOnlyLatestState = true
+		c.Ledger.RemoveUntraceableBlocks = true
+	}
+	storageCfg := func(c *config.Blockchain) {
+		spoutCfg(c)
+		c.Ledger.KeepOnlyLatestState = true
+		c.Ledger.RemoveUntraceableBlocks = true
+		c.NeoFSStateSyncExtensions = true
+		c.NeoFSStateFetcher.Enabled = true
+		c.NeoFSBlockFetcher.Enabled = true
+	}
+
+	cases := []struct {
+		name string
+		cfg  func(*config.Blockchain)
+	}{
+		{"MPT-based", mptCfg},
+		{"storage-based", storageCfg},
+	}
+
+	for _, cc := range cases {
+		t.Run(cc.name, func(t *testing.T) {
+			bcBolt, _, _ := chain.NewMultiWithCustomConfig(t, cc.cfg)
+			module := bcBolt.GetStateSyncModule()
+
+			var calls int
+			module.SetOnStageChanged(func() { calls++ })
+
+			require.NoError(t, module.Init(bcSpout.BlockHeight()))
+			require.Equal(t, 1, calls)
+
+			syncPoint := module.GetStateSyncPoint()
+			// AddHeaders up to P → headersSynced → active
+			for i := uint32(1); i <= bcSpout.HeaderHeight(); i++ {
+				h, err := bcSpout.GetHeader(bcSpout.GetHeaderHash(i))
+				require.NoError(t, err)
+				require.NoError(t, module.AddHeaders(h))
+				if i > syncPoint {
+					break
+				}
+			}
+			require.Equal(t, 2, calls)
+			require.True(t, module.IsActive())
+			// AddMPTNodes or AddContractStorageItems up to P → storageSynced
+			if cc.name == "MPT-based" {
```

Comment:
It's a bad pattern, don't bind it to test name. Pass mode as a parameter to checking function, see how it is done in TestStateSyncModule_RestoreBasicChain for example.